http-server/src/main/java/org/triplea/server/moderator/toolbox/api/key/KeyHasher.java

```diff
@@ -0,0 +1,27 @@
+package org.triplea.server.moderator.toolbox.api.key;
+
+import com.google.common.base.Charsets;
+import com.google.common.base.Preconditions;
+import com.google.common.hash.Hashing;
+
+import lombok.AccessLevel;
+import lombok.NoArgsConstructor;
+
+/**
+ * Utility class to provide hashing functions for moderator API key.
+ */
+@NoArgsConstructor(access = AccessLevel.PRIVATE)
+public final class KeyHasher {
+
+  public static String applyHash(final String apiKey, final String password) {
+    Preconditions.checkNotNull(apiKey);
+    Preconditions.checkNotNull(password);
+    return applyHash(apiKey + password);
+  }
+
+  public static String applyHash(final String valueToHash) {
+    Preconditions.checkNotNull(valueToHash);
+    Preconditions.checkArgument(!valueToHash.isEmpty());
+    return Hashing.sha512().hashString(valueToHash, Charsets.UTF_8).toString();
```

Comment:
I like the abstraction, but at least for the engine code SHA hashes should have inconsistent casing, so it might be worth using a consistent case to avoid confusion in case we have to deal with the actual hashes at some point.

Reply:
Sorry, I don't quite follow "actual hashes" and consistent vs inconsistent casing.

Reply:
IIRC,in the game engine, sha-512 always looks like this: aaabbeeff13477
Whereas the api you're using might be outputting this: AABBEEFF13477

Reply:
Sorry @RoiEXLab , I don't understand either the concern or the suggested course of action.

> the engine code SHA hashes should have inconsistent casing

You desire an inconsistent casing?

>  it might be worth using a consistent case

Are we shooting for an inconsistent casing? I'm confused.
Do you have suggestions for how to get an inconsistent casing if so?

>  avoid confusion in case we have to deal with the actual hashes at some point.

I'm not sure which 'actual hashes' are being referenced here.

Perhaps there is missing context, we are hashing a brand new key, the hashed key does not interact with the game engine. In the case of a new API key, generated on the server, the unhashed value is sent back to the client and that same value is returned to the server. All hashing is done by the server.

--------------

FWIW, it looks like the `Hashing.sha512` function produces lower case output. If the suggestion is to `toLowerCase()` the value, I'm not sure if that is a good move:
- looks like it is redundant, be what it may
- constraining the output could increase the chance of hash collisions. It seems like the output hash is complex anyways to avoid that

Reply:
Ok, note to myself: Do not reply in a hurry when just out of bed 😅 
What I was trying to say is:
I don't know the Hashing class. The java standard library does hashing all lowercase, where as the guava tooling IIRC does hashing all upper case.
I just wanted to be sure that we don't end up comparing the same hash with a different case at some point, just because of some random implementation difference like this. Might be worth documenting if the output does interact with user input at some point.

Off Topic:
It's interesting to see how language differences (german and english in particular) affect the desired intention of a sentence.
> the engine code SHA hashes should have inconsistent casing

while this sentence is clearly wrong (I don't even know why I wrote this, it makes 0 sense 😅), the "should" in this sentence was supposed to mean "should" as in "to my knowledge I expect X to happen" rather than "in an ideal world X is the desired outcome".

I honestly don't know if both meanings are equally valid in english, in german they definitely are.


Anyways I hope everything should be clear now ^^

Reply:
FWIW:

> the engine code SHA hashes should have inconsistent casing

Rephrasing with more idiomatic english:
- "the engine code SHA hashes I believe have inconsistent casing"
- "the engine code SHA hashes are an inconsistent casing"

For expressing uncertainty and also conjecturing:
- "I thought the engine code SHA hashes would have an inconsistent casing"
- "I thought they would have an inconsistent casing"

Reply:
Guava hashing sha-512 appears to be all lower case.

Doing some quick reading on this, we probably want to use bcrypt. A fast algorithm like sha-512 is not that great, if DB data were leaked then it would not take too long to start cracking the hashes.

Reply:
I added a TODO comment to note that we should use bcrypt: 803ad49b8

Reply:
Fair point. The API key is somewhat of a password replacement, but hashing isn't as crucial because they are long enough completely random tokens in a ideal world.
Adding that extra bit of security is a good idea nevertheless.